packages/graphql-server/src/plugins/useRedwoodDirective.ts

```diff
@@ -91,79 +95,129 @@ export function hasDirective(info: GraphQLResolveInfo): boolean {
+              return originalResolve(root, args, context, info)
+            },
+          }
         }
-
-        // In order to change the value of the field, we have to return a function in this form
-        // ({result, setResult}) => { setResult(newValue)}
-        // Not super clear but mentioned here: https://www.envelop.dev/docs/plugins/lifecycle#onexecuteapi
-
         if (_isTransformer(options)) {
-          return ({ result, setResult }) => {
-            // @NOTE! A transformer cannot be async
-            const transformedValue = options.onResolverCalled({
+          return {
+            ...fieldConfig,
+            resolve: function useRedwoodDirectiveTransformerResolver(
               root,
               args,
               context,
-              info,
-              directiveNode,
-              directiveArgs,
-              resolvedValue: result,
-            })
-
-            setResult(transformedValue)
+              info
+            ) {
+              const resolvedValue = originalResolve(root, args, context, info)
+              if (isPromise(resolvedValue)) {
+                return resolvedValue.then((resolvedValue) =>
+                  options.onResolverCalled({
+                    root,
+                    args,
+                    context,
+                    info,
+                    directiveNode,
+                    directiveArgs,
+                    resolvedValue,
+                  })
+                )
+              }
+              return options.onResolverCalled({
+                root,
+                args,
+                context,
+                info,
+                directiveNode,
+                directiveArgs,
+                resolvedValue,
+              })
+            },
           }
         }
       }
+      return fieldConfig
+    },
+  })
+}
 
-      return
+export const useRedwoodDirective = (
+  options: DirectivePluginOptions
+): Plugin<{
+  onResolverCalled: ValidatorDirectiveFunc | TransformerDirectiveFunc
+}> => {
+  /**
+   * This symbol is added to the schema extensions for checking whether the transform got already applied.
+   */
+  const didMapSchemaSymbol = Symbol('useRedwoodDirective.didMapSchemaSymbol')
+  return {
+    onSchemaChange({ schema, replaceSchema }) {
+      /**
+       * Currently graphql-js extensions typings are limited to string keys.
+       * We are using symbols as each useRedwoodDirective plugin instance should use its own unique symbol.
+       */
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      if (schema.extensions?.[didMapSchemaSymbol] === true) {
+        return
+      }
+      const transformedSchema = wrapAffectedResolvers(schema, options)
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      transformedSchema.extensions = {
+        ...schema.extensions,
+        [didMapSchemaSymbol]: true,
+      }
+      replaceSchema(transformedSchema)
```

Comment:
Since forevery replaceSchema is called the onSchemaChange hook of all other plugins is called we need a way to track whether a transform has already been applied for preventing an infinite loop. The schema extensions are perfect for this!